### ISSUE TITLE
Populating virtuals with count: true fires a 'find' query instead of 'countDocuments' query

### ISSUE THREAD

**Issue by abhishekjain2604** [Automattic/mongoose#8121]
**Do you want to request a *feature* or report a *bug*?**
I want to suggest an optimisation for an existing feature.

**What is the current behavior?**
Here is the script that I'm using
```js

const mongoose = require('mongoose');
mongoose.Promise = global.Promise;

mongoose.connect('mongodb://localhost:27017', { dbName: 'test-db', useNewUrlParser: true }, function (error) {
  if (error) {
    console.log('Error in mongo connection', error);
  } else {
    console.log('Mongo connection established');
  }
});
mongoose.set('debug', true);

const PersonSchema = new mongoose.Schema({
  name: String,
  band: String,
});
let modelPerson = mongoose.model('person', PersonSchema);


const BandSchema = new mongoose.Schema({
  name: String
}, { toJSON: { virtuals: true } });
BandSchema.virtual('numMembers', {
  ref: 'person',
  localField: 'name',
  foreignField: 'band',
  count: true
});
let modelBand = mongoose.model('band', BandSchema);

let main = async () => {
  await modelPerson.insertMany([{ name: 'P1', band: 'B1' }, { name: 'P2', band: 'B1' }]);
  await modelBand.create({ name: 'B1' });

  let band = await modelBand.findOne({ name: 'B1' }).populate('numMembers');
  console.log(band.numMembers);

};

main();
```

In the findOne query for modelBand, populating 'numMembers' fires a find query on modelPerson instead of a simple countDocuments query which is hghly inefficient considering the count can go well above 1000 in many cases. Actually fetching all those documents will reduce the response time and consume more bandwidth.

In this case, since all we really need is a count of those documents, a countDocuments query is a perfect fit here.

**What are the versions of Node.js, Mongoose and MongoDB you are using? Note that "latest" is not a version.**
Node.js: 8.10
MongoDb: 4.0.4
Mongoose: 5.6.11

**Comment by vkarpov15** [comment]
Your suggestion is a potential optimization in the case of `findOne()`, but the reason why we need `find()` instead of `countDocuments()` is this: suppose you did `modelBand.find()` and there were multiple bands in the database. One `countDocuments()` query wouldn't work, because there would be no way to figure out what `numMembers` should be for each individual band.

**Comment by abhishekjain2604** [comment]
Makes sense. 

But I think even in case of `find()` we can still limit the number of fields that are fetched for the documents using projections and restrict it to just the `_id` and the localField that was set when defining virtuals

**Comment by abhishekjain2604** [comment]
Also, another alternative can be an aggregate query where we can fetch the counts that are already groups by the `foreignField` of the modelPerson.

Would that help? Performance wise?

**Comment by vkarpov15** [comment]
@abhishekjain2604 we already limit the fields - do you have a case where populate with count selects all fields?

And I like the aggregation idea, that would probably be more efficient. Thanks for the suggestion, we will look into it.

**Comment by abhishekjain2604** [comment]
@vkarpov15 Yes the fields are indeed being limited using projections. My bad
